neurom/core/features.py

```diff
@@ -0,0 +1,136 @@
+# All rights reserved.
+#
+# This file is part of NeuroM <https://github.com/BlueBrain/NeuroM>
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+#     1. Redistributions of source code must retain the above copyright
+#        notice, this list of conditions and the following disclaimer.
+#     2. Redistributions in binary form must reproduce the above copyright
+#        notice, this list of conditions and the following disclaimer in the
+#        documentation and/or other materials provided with the distribution.
+#     3. Neither the name of the copyright holder nor the names of
+#        its contributors may be used to endorse or promote products
+#        derived from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY
+# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+'''Feature registrationg and retrieval'''
+
+import numpy as np
+
+from functools import partial
+from ..exceptions import NeuroMError
+from .types import tree_type_checker
+from . import iter_neurites
+from . import NeuriteType
+
+
+FEATURES = {'NEURITEFEATURES': {},
+            'NEURONFEATURES': {}, }
+
+
+def register_neurite_feature(name, func):
+    '''Register a feature to be applied to neurites
+
+    Parameters:
+        name: name of the feature, used for access via get() function.
+        func: single parameter function of a neurite.
+    '''
+    def _fun(neurites, neurite_type=NeuriteType.all):
+        '''Wrap neurite function from outer scope and map into list'''
+        return list(func(n)
+                    for n in iter_neurites(neurites, filt=tree_type_checker(neurite_type)))
+
+    register_feature('NEURITEFEATURES', name, _fun)
+
+
+def register_feature(namespace, name, func):
+    '''Register a feature to be applied
+
+    Parameters:
+        namespace(string):
+        name(string): name of the feature, used for access via get() function.
+        func(callable): single parameter function of a neurite.
+    '''
+    if name in FEATURES[namespace]:
+        raise NeuroMError('Attempt to hide registered feature %s' % name)
+
+    FEATURES[namespace][name] = func
+
+
+def feature(func=None, namespace=None, names=''):
```

Comment:
parameter name 'aliases' would convey the meaning better.
And why is it a space-separated string rather than list?